### PR TITLE
Fixes some error handling when unexpected input is given

### PR DESCRIPTION
**Pull request by schwartzmx**
The current `stackit`, throws some nasty stack trace if invalid input is given after searching for a question.  See:

```
Press m for more, or a number to select: s
Traceback (most recent call last):
  File "/usr/local/bin/stackit", line 9, in <module>
    load_entry_point('stackit==0.1.2', 'console_scripts', 'stackit')()
  File "/Library/Python/2.7/site-packages/stackit/stackit_core.py", line 188, in main
    searchTerm(term, tags)
  File "/Library/Python/2.7/site-packages/stackit/stackit_core.py", line 76, in searchTerm
    focusQuestion(questionLogs, count)
  File "/Library/Python/2.7/site-packages/stackit/stackit_core.py", line 43, in focusQuestion
    if(0 < int(userInput) and int(userInput) <= count):
ValueError: invalid literal for int() with base 10: 's'
```

I altered some wording, and also added some error catching to make it not spit out the stack trace.  Also added `q` option for quitting a search.

Comments or concerns let me know.

Thanks,

Phil

### CODE REVIEW

**Comment by lukasschwab** [comment]
Ayyy Phil

We dig it––thanks for catching the issue and implementing the "quit search" feature, which we've wanted to implement but haven't gotten around to doing.

We'll test your code and get back to you, hopefully later today!

Thanks for the pull request,
Lukas

File: stackit/stackit_core.py [hunk 1]
```diff
@@ -39,24 +39,31 @@ def focusQuestion(questions, count):
...
+                        webbrowser.open(questions[int(userInput)-1].json['link'], new=0, autoraise=True)
+                #User selects x to return to search
+                if(branchInput == 'x'):
+                    print("\n\n\n\n\n\n\n\n\n\n\n\n")
+                    #Ranging over the 5 questions including the user's choice
+                    for j in range(5*int((int(userInput)-1)/5), 5*int((int(userInput)-1)/5)+5):
+                        printQuestion(questions[j], j+1)
+                    continue   #exit the inner while loop
+        except:
+            if (userInput != 'q'):
```
Comment (lukasschwab) [inline stackit/stackit_core.py:80]: The `userInput == 'q'` case is handled above, in line 44––is there a reason for this redundancy?

File: stackit/stackit_core.py [hunk 2]
```diff
@@ -39,24 +39,31 @@ def focusQuestion(questions, count):
...
+                print("The input entered was not recognized as a valid choice.  Exiting...")
```
Comment (lukasschwab) [inline stackit/stackit_core.py]: Though there's a possibility that this try/except structure could throw errors besides having an invalid character, that seems unlikely.

Why is this "invalid input" warning not printed every time an error is thrown in this block?

File: stackit/stackit_core.py [hunk 3]
```diff
@@ -39,24 +39,31 @@ def focusQuestion(questions, count):
...
+            sys.exit()
```
Comment (lukasschwab) [inline stackit/stackit_core.py]: Would rather this point returned to the initial prompt––"Enter m for more, a question number to select, or q to exit: "––instead of sys.exit()-ing. Allows for user mistakes. There is a manual sys.exit() option in the input 'q' now, as well as control-C.

tl;dr: would rather not quit without explicit user request.

File: stackit/stackit_core.py (hunk 1 above)
Comment (schwartzmx) [inline stackit/stackit_core.py:80]: When you throw the exception after the check, it automatically propagates to the `except:` block.  So when letting the user know what happened, you need to check if the exception occured because of `q`, if it didn't then let the user know why. If it did then just exit as normal.

File: stackit/stackit_core.py (hunk 2 above)
Comment (schwartzmx) [inline stackit/stackit_core.py]: It is printed everytime, only not printed when the choice `q` is entered...

Comment (lukasschwab) [inline stackit/stackit_core.py]: Gotcha––that makes sense. I wasn't thinking of `sys.exit()` as throwing an exception, but this is definitely cleared up.

**Comment by schwartzmx** [comment]
I also added similar checking for `branchInput` looping, because the same error could occur.  Also if a user found the answer after taking a closer look (i.e. by entering a question number), give them the option to quit right there, and not have to `ctrl-c` or hit `x` then `q`.  

Also give a meaningful message on why nothing happens when an integer out of the range of `(0 < int(userInput) and int(userInput) <= count)` is entered... and then re-prompt.

Let me know if you have anymore concerns.

Thanks,

Phil

File: stackit/stackit_core.py (hunk 3 above)
Comment (schwartzmx) [inline stackit/stackit_core.py]: So would you rather it be `continue` then `sys.exit()`?

Comment (lukasschwab) [inline stackit/stackit_core.py]: I think I'd rather it was `continue` than `sys.exit()` in the case that
some invalid character is entered. e.g.:

    except:
        if (userInput != 'q'):
            print("The input entered was not recognizes as a valid choice.")
            continue
        else: # If the user input is 'q', no input
            sys.exit()

On Wed, Feb 4, 2015 at 12:55 PM, Phil <notifications@github.com> wrote:

> In stackit/stackit_core.py
> <https://github.com/lukasschwab/stackit/pull/4#discussion_r24118595>:
>
> > +                branchInput = '0'   #user deciding whether to branch to open browser, or to return to search
> > +                while(branchInput != 'x'):
> > +                    branchInput = promptUser("Enter b to launch browser, or x to return to search: ")
> > +                    if(branchInput == 'b'):
> > +                        webbrowser.open(questions[int(userInput)-1].json['link'], new=0, autoraise=True)
> > +                #User selects x to return to search
> > +                if(branchInput == 'x'):
> > +                    print("\n\n\n\n\n\n\n\n\n\n\n\n")
> > +                    #Ranging over the 5 questions including the user's choice
> > +                    for j in range(5*int((int(userInput)-1)/5), 5*int((int(userInput)-1)/5)+5):
> > +                        printQuestion(questions[j], j+1)
> > +                    continue   #exit the inner while loop
> > +        except:
> > +            if (userInput != 'q'):
> > +                print("The input entered was not recognized as a valid choice.  Exiting...")
> > +            sys.exit()
>
> So would you rather it be continue then sys.exit()?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lukasschwab/stackit/pull/4/files#r24118595>.
>



-- 
Lukas Schwab

Comment (lukasschwab) [inline stackit/stackit_core.py]: I think I'd rather it was `continue` than `sys.exit()` in the case that
some invalid character is entered. e.g.:

``` python
except:
    if (userInput != 'q'):
        print("The input entered was not recognizes as a valid choice.")
        continue
    else: # If the user input is 'q', no warning is necessary
        sys.exit()
```

Comment (schwartzmx) [inline stackit/stackit_core.py]: Got it, added in latest commit, works for both `userInput` and `branchInput`.  It does make more sense.  So now it prompts the user to re-enter, while also letting them know it was invalid input.
  :+1: 

```
Enter b to launch browser, x to return to search, or q to exit: s
The input entered was not recognized as a valid choice.
Enter b to launch browser, x to return to search, or q to exit:
```

**Comment by lukasschwab** [comment]
@schwartzmx our first pull request is now merged!

For now, you will be listed in README.md as Phil, with a link to your Github profile––let me know if you want me to display any more//less information.

Thanks for working with us!

**Comment by schwartzmx** [comment]
@lukasschwab No problem man, and what a great idea you all came up with.  As I use it more and find other features that may be useful I won't hesitate to open another PR.  

Edit: And thats fine for the `README`. Thanks

Cheers, have a good night.

Phil
